src/tree_gui.cpp

```diff
@@ -41,134 +45,238 @@ const PalSpriteID tree_sprites[] = {
+		}
+	}
 
-		int i = this->base + widget - WID_BT_TYPE_11;
-		/* Trees "grow" in the centre on the bottom line of the buttons */
-		DrawSprite(tree_sprites[i].sprite, tree_sprites[i].pal, (r.left + r.right) / 2 + WD_FRAMERECT_LEFT, r.bottom - 7);
+	void DrawWidget(const Rect &r, int widget) const override
+	{
+		if (widget >= WID_BT_TYPE_BUTTON_FIRST) {
+			const int index = widget - WID_BT_TYPE_BUTTON_FIRST;
+			/* Trees "grow" in the centre on the bottom line of the buttons */
+			DrawSprite(tree_sprites[index].sprite, tree_sprites[index].pal, (r.left + r.right) / 2 + WD_FRAMERECT_LEFT, r.bottom - ScaleGUITrad(BUTTON_BOTTOM_OFFSET));
+		}
 	}
 
 	void OnClick(Point pt, int widget, int click_count) override
 	{
 		switch (widget) {
-			case WID_BT_TYPE_11: case WID_BT_TYPE_12: case WID_BT_TYPE_13: case WID_BT_TYPE_14:
-			case WID_BT_TYPE_21: case WID_BT_TYPE_22: case WID_BT_TYPE_23: case WID_BT_TYPE_24:
-			case WID_BT_TYPE_31: case WID_BT_TYPE_32: case WID_BT_TYPE_33: case WID_BT_TYPE_34:
-				if (widget - WID_BT_TYPE_11 >= this->count) break;
-
-				if (HandlePlacePushButton(this, widget, SPR_CURSOR_TREE, HT_RECT)) {
-					this->tree_to_plant = (TreeType)(this->base + widget - WID_BT_TYPE_11);
-				}
-				break;
-
 			case WID_BT_TYPE_RANDOM: // tree of random type.
-				if (HandlePlacePushButton(this, WID_BT_TYPE_RANDOM, SPR_CURSOR_TREE, HT_RECT)) {
-					this->tree_to_plant = TREE_INVALID;
-				}
+				this->tree_to_plant = this->tree_to_plant == TREE_INVALID ? -1 : TREE_INVALID;
+				this->UpdateMode();
 				break;
 
 			case WID_BT_MANY_RANDOM: // place trees randomly over the landscape
 				if (_settings_client.sound.confirm) SndPlayFx(SND_15_BEEP);
 				PlaceTreesRandomly();
 				MarkWholeScreenDirty();
 				break;
+
+			case WID_BT_MODE_NORMAL:
+				this->mode = PM_NORMAL;
+				this->UpdateMode();
+				break;
+			case WID_BT_MODE_FOREST_SM:
+				assert(_game_mode = GM_EDITOR);
+				this->mode = PM_FOREST_SM;
+				this->UpdateMode();
+				break;
+			case WID_BT_MODE_FOREST_LG:
+				assert(_game_mode = GM_EDITOR);
+				this->mode = PM_FOREST_LG;
+				this->UpdateMode();
+				break;
+
+			default:
+				if (widget >= WID_BT_TYPE_BUTTON_FIRST) {
+					const int index = widget - WID_BT_TYPE_BUTTON_FIRST;
+					this->tree_to_plant = this->tree_to_plant == index ? -1 : index;
+					this->UpdateMode();
+				}
+				break;
 		}
 	}
 
 	void OnPlaceObject(Point pt, TileIndex tile) override
 	{
-		VpStartPlaceSizing(tile, VPM_X_AND_Y, DDSP_PLANT_TREES);
+		if (this->mode == PM_NORMAL) {
+			VpStartPlaceSizing(tile, VPM_X_AND_Y, DDSP_PLANT_TREES);
+		} else {
+			this->DoPlantForest(tile);
+		}
 	}
 
 	void OnPlaceDrag(ViewportPlaceMethod select_method, ViewportDragDropSelectionProcess select_proc, Point pt) override
 	{
-		VpSelectTilesWithMethod(pt.x, pt.y, select_method);
+		if (this->mode == PM_NORMAL) {
+			VpSelectTilesWithMethod(pt.x, pt.y, select_method);
+		}
```

Comment:
missing DoPlantForest in this function? Can't click & drag for grove/forest types